File: client/ios/DivKit/Actions/TooltipActionHandler.swift
import LayoutKit

final class TooltipActionHandler {
  private let performer: TooltipActionPerformer?
  private let reporter: DivReporter

  init(
    performer: TooltipActionPerformer?,
    reporter: DivReporter
  ) {
    self.performer = performer
    self.reporter = reporter
  }

  func handle(_ action: DivActionShowTooltip, context: DivActionHandlingContext) {
    guard let id = action.resolveId(context.expressionResolver),
          let multiple = action.resolveMultiple(context.expressionResolver) else {
      return
    }

    showTooltip(makeTooltipInfo(id: id, multiple: multiple, context: context))
  }

  func handle(_ action: DivActionHideTooltip, context: DivActionHandlingContext) {
    guard let id = action.resolveId(context.expressionResolver) else {
      return
    }

    hideTooltip(
      TooltipIdentity(id: id, scopePath: context.scopePath)
    )
  }

  func handleShowTooltip(id: String, multiple: Bool, context: DivActionHandlingContext) {
    showTooltip(makeTooltipInfo(id: id, multiple: multiple, context: context))
  }

  func handleHideTooltip(id: String, context: DivActionHandlingContext) {
    hideTooltip(
      TooltipIdentity(id: id, scopePath: context.scopePath)
    )
  }

  func showTooltip(_ info: TooltipInfo) {
    performer?.showTooltip(info: info)
  }

  func hideTooltip(_ identity: TooltipIdentity) {
    performer?.hideTooltip(identity: identity)
  }

  private func makeTooltipInfo(
    id: String,
    multiple: Bool,
    context: DivActionHandlingContext
  ) -> TooltipInfo {
    let cardId = context.cardId
    let sourcePath = context.sourcePath
    return TooltipInfo(
      id: id,
      showsOnStart: false,
      multiple: multiple,
      scopePath: context.scopePath,
      onError: { [weak self] message in
        self?.reporter.reportError(
          cardId: cardId,
          error: DivUnknownError(message: message, path: sourcePath)
        )
      }
    )
  }
}

File: client/android/compose/src/main/kotlin/com/yandex/div/compose/actions/TooltipActionHandler.kt
package com.yandex.div.compose.actions

import com.yandex.div.compose.tooltips.TooltipStateStorage
import com.yandex.div.internal.actions.DivUntypedAction
import com.yandex.div2.DivActionHideTooltip
import com.yandex.div2.DivActionShowTooltip
import javax.inject.Inject

internal class TooltipActionHandler @Inject constructor(
    private val tooltipStateStorage: TooltipStateStorage
) {

    fun handle(context: DivActionHandlingContext, action: DivActionShowTooltip) {
        val id = action.id.evaluate(context.expressionResolver)
        tooltipStateStorage.show(id)
    }

    fun handle(context: DivActionHandlingContext, action: DivActionHideTooltip) {
        val id = action.id.evaluate(context.expressionResolver)
        tooltipStateStorage.hide(id)
    }

    fun handle(action: DivUntypedAction.ShowTooltip) {
        tooltipStateStorage.show(action.id)
    }

    fun handle(action: DivUntypedAction.HideTooltip) {
        tooltipStateStorage.hide(action.id)
    }
}
